minetest/minetest
# [Manual Squash] Redesign/unify mainmenu settings interface

Fixes #12452, fixes #3266, fixes #12976, fixes #9016, fixes #12140, fixes #11069

The settings tab and all settings has been replaced with a single interface for changing settings

![image](https://user-images.githubusercontent.com/2122943/232167857-2baac6f6-c38c-431b-a464-20e8ee50cdc5.png)

## Technical intro

The dialog has a list of pages grouped into sections. When the dialog loads, it builds the pages by reading settingtypes.txt. 

When opening a page, the dialog builds components for each setting in the page. It's possible to define custom components in the setting menu, this is done to retain the current logic for the shader selection component.

A component is a table with a get_formspec function. Functions in components.lua build components for settings.

## To do

This PR is Ready for Review

- [x] Reimplement
- [x] Use scrollbar options to get proper scrollbars
- [x] Readd key changes
- [x] Show current page in leftpane
- [x] Restructure settingtypes.txt
- [x] Fix translation
- [x] Filter pages when searching
- [x] Validate min max
- [x] Readd reset
- [x] Restore element behaviour on "Most used" pane (what was the settings tab, eg: the shadow quality dropdown)
- [x] Readd show_technical_names
- [x] Fix issue with dropdown events (see todo in bool component)
- [x] Undersampling defaults to 0 in .cpp, but 1 in settingtypes.txt
- [x] Finish support for path select dialog
- [x] Improve styling 
- [x] The "Edit" button for the mapgen noises should be a bit wider (in German, the translation barely fits)
- [x] `1 ~= 1.0` reset button bug
- [x] Use image and tooltip area for info icons
- [x] Info icon for noise settings
- [x] Crash when editing a text field, eg: selectionbox_color
- [x] Icon still appears when there's no description, eg: font_bold
- [x] Larger box around show_technical_names
- [x] ~~Replace dummy tab with settings icon~~ (unnecessary work)
- [x] ~~Most Used > shader settings: Show technical names~~ (planning to remove Most Used in the future)


For future PRs (in rough order of priority):

- Fix dropdowns changing when scrolling ( https://github.com/minetest/irrlicht/pull/192 )
- Only show settings when relevant (ie: don't show shader settings when shaders are disabled), remove Most Used tab
- Add flag utils to Settings ref
- Sort mods based on title when show_technical_names=false
- Hide advanced settings by default (including mapgen settings)
- Noise dialog needs validation for Nan, inf, -inf.
- Noise dialog needs tooltips
- Support reset for noise parameters
- Improve keyboard use
- Mapgen flags translation
- Better use of font size (eg: of sub section headers)
- Support for drop down translation (hard coded)
- Add color type
- Use slider for some fields (like mouse sensitivity)
- Better way to show unchanged formspec ( https://github.com/minetest/minetest/issues/10432 )
- Limit font size / support larger fonts

## How to test

Types of setting: int, float, string, bool, v3f, flags, path, filepath, noise, enum

* Try editing one of each type. Make sure reset icon appears when relevant
* Try resetting one of each type
* Browse through, checking layout/structure/etc
